Source/Core/DolphinQt/Achievements/AchievementProgressWidget.cpp

```diff
@@ -42,33 +42,47 @@ void AchievementProgressWidget::UpdateData(bool clean_all)
   {
     m_achievement_boxes.clear();
     ClearLayoutRecursively(m_common_layout);
-
-    auto& instance = AchievementManager::GetInstance();
-    if (!instance.IsGameLoaded())
-      return;
-    auto* client = instance.GetClient();
-    auto* achievement_list = rc_client_create_achievement_list(
-        client, RC_CLIENT_ACHIEVEMENT_CATEGORY_CORE_AND_UNOFFICIAL,
-        RC_CLIENT_ACHIEVEMENT_LIST_GROUPING_PROGRESS);
-    for (u32 ix = 0; ix < achievement_list->num_buckets; ix++)
+  }
+  else
+  {
+    while (auto* item = m_common_layout->takeAt(0))
     {
-      m_common_layout->addWidget(new QLabel(tr(achievement_list->buckets[ix].label)));
-      for (u32 jx = 0; jx < achievement_list->buckets[ix].num_achievements; jx++)
+      auto* widget = item->widget();
+      m_common_layout->removeWidget(widget);
+      if (std::strcmp(widget->metaObject()->className(), "QLabel") == 0)
       {
-        auto* achievement = achievement_list->buckets[ix].achievements[jx];
-        m_achievement_boxes[achievement->id] = std::make_shared<AchievementBox>(this, achievement);
-        m_common_layout->addWidget(m_achievement_boxes[achievement->id].get());
+        delete widget;
+        delete item;
       }
     }
-    rc_client_destroy_achievement_list(achievement_list);
   }
-  else
+
+  auto& instance = AchievementManager::GetInstance();
+  if (!instance.IsGameLoaded())
+    return;
+  auto* client = instance.GetClient();
+  auto* achievement_list =
+      rc_client_create_achievement_list(client, RC_CLIENT_ACHIEVEMENT_CATEGORY_CORE_AND_UNOFFICIAL,
+                                        RC_CLIENT_ACHIEVEMENT_LIST_GROUPING_PROGRESS);
+  for (u32 ix = 0; ix < achievement_list->num_buckets; ix++)
   {
-    for (auto box : m_achievement_boxes)
+    m_common_layout->addWidget(new QLabel(tr(achievement_list->buckets[ix].label)));
+    for (u32 jx = 0; jx < achievement_list->buckets[ix].num_achievements; jx++)
     {
-      box.second->UpdateData();
+      auto* achievement = achievement_list->buckets[ix].achievements[jx];
+      auto box_itr = m_achievement_boxes.find(achievement->id);
```

Comment:
In the case of not finding the id, you are doing another lookup, consider instead to use a hint:

```
auto box_itr = m_achievement_boxes.lower_bound(achievement->id);
if (box_itr != m_achievement_boxes.end() && box_itr->first == achievement->id)
{
  box_itr->second->UpdateProgress();
}
else
{
  m_achievement_boxes.try_emplace(box_itr, achievement->id, std::make_shared<AchievementBox>(this, achievement));
}
 ```

Reply:
And actually noticed there's _another_ lookup when `addWidget` is called, we can fix that..

```
auto box_itr = m_achievement_boxes.lower_bound(achievement->id);
if (box_itr != m_achievement_boxes.end() && box_itr->first == achievement->id)
{
  box_itr->second->UpdateProgress();
  m_common_layout->addWidget(box_itr->second.get());
}
else
{
  const auto new_box_itr = m_achievement_boxes.try_emplace(box_itr, achievement->id, std::make_shared<AchievementBox>(this, achievement));
  m_common_layout->addWidget(new_box_itr->second.get());
}
```